skopt/tests/test_utils.py

```diff
@@ -165,3 +169,59 @@ def test_normalize_dimensions(dimensions, normalizations):
     space = normalize_dimensions(dimensions)
     for dimension, normalization in zip(space, normalizations):
         assert dimension.transform_ == normalization
+
+
+@pytest.mark.fast_test
+def test_use_named_args():
+    """
+    Test the function wrapper @use_named_args which is used
+    for wrapping an objective function with named args so it
+    can be called by the optimizers which only pass a single
+    list as the arg.
+    
+    This test does not actually use the optimizers but merely
+    simulates how they would call the function.
+    """
+
+    # Define the search-space dimensions. They must all have names!
+    dim1 = Real(name='foo', low=0.0, high=1.0)
+    dim2 = Real(name='bar', low=0.0, high=1.0)
+    dim3 = Real(name='baz', low=0.0, high=1.0)
+
+    # Gather the search-space dimensions in a list.
+    dimensions = [dim1, dim2, dim3]
+
+    # Parameters that will be passed to the objective function.
+    default_parameters = [0.5, 0.6, 0.8]
+
+    # Define the objective function with named arguments
+    # and use this function-decorator to specify the search-space dimensions.
+    @use_named_args(dimensions=dimensions)
+    def func(foo, bar, baz):
+        # Assert that all the named args are indeed correct.
+        assert foo == default_parameters[0]
+        assert bar == default_parameters[1]
+        assert baz == default_parameters[2]
+
+        # Return some objective value.
+        return foo ** 2 + bar ** 4 + baz ** 8
+
+    # Ensure the objective function can be called with a single
+    # argument named x.
+    res = func(x=default_parameters)
+    assert (isinstance(res, float))
+
+    # Ensure the objective function can be called with a single
+    # argument that is unnamed.
+    res = func(default_parameters)
+    assert (isinstance(res, float))
+
+    # Ensure the objective function can be called with a single
+    # argument that is a numpy array named x.
+    res = func(x=np.array(default_parameters))
+    assert (isinstance(res, float))
+
+    # Ensure the objective function can be called with a single
+    # argument that is an unnamed numpy array.
+    res = func(np.array(default_parameters))
+    assert (isinstance(res, float))
```

Comment:
Maybe add checks whether with unnamed dimensions / non - `Dimension` objects an exception rises. I imagine that in future we would want to extend your code to support also implicit dimension definitions with names, which could look like:
```python
(0.1, 1.0, 'log-uniform', 'foo')  # equivalent to Real(0.1, 1.0, prior='log-uniform', name='foo')
# currently we can only use
# (0.1, 1.0, 'log-uniform') 
# and then the dimension is unnamed
```
which might temper with these exceptions, and could lead to subtle bugs.  While such definition of dimension seems to be somewhat ugly at first glance, it is appealing in that it leads to a more compact code.

Reply:
(in case you need further info on various dimension definitions, see [here](https://github.com/scikit-optimize/scikit-optimize/blob/master/examples/hyperparameter-optimization.ipynb) or [here](https://github.com/scikit-optimize/scikit-optimize/blob/master/examples/sklearn-gridsearchcv-replacement.ipynb))

Reply:
Hmm ... I have tested myself whether the exceptions are thrown and the new detailed error-messages are correct. I'm not sure a unit-test adds much "safety" to this? And I don't quite understand what subtle bugs might arise in the future?

Furthermore, regarding your short-hand notation I think it is actually quite problematic, both in terms of reading your code that uses skopt, such as this:

    space  = [(1, 5),                           # max_depth
              (10**-5, 10**0, "log-uniform"),   # learning_rate
              (1, n_features),                  # max_features
              (2, 100),                         # min_samples_split
              (1, 100)]                         # min_samples_leaf

In my opinion, it would be much more clear if you wrote:

    space  = [Int(1, 5, name='max_depth'),
              Int(10**-5, 10**0, prior="log-uniform", name='learning_rate'),
              Int(1, n_features, name='max_features'),
              Int(2, 100, name='min_samples_split'), 
              Int(1, 100, name='min_samples_leaf')] 

I can also see in your code that the short-hand notation causes problems other places and it requires numerous weird tests, etc.

You will need even more strange tests if you allow naming in the short-hand notation, because now you don't know whether the 3rd arg is the name of the dimension, or if it defines e.g. the prior-mapping. You can fix that with more weird tests, such as whether the 3rd string arg is one of the allowed prior-mappings, otherwise it is assumed to be the name. But what if the user has made a typo in the prior-mapping? ... Suddenly the API and code is very confusing and bloated. It's not going to be pretty! :-)

In my opinion, it is better to aim at making the code clear even if it means it becomes significantly more verbose. Having compactness of the code as a major goal is only really justified in performance-critical applications - in which case the code can be so hard to read, that extensive commenting is then needed.

Reply:
Yeah I also expect the short-hand definition of dimension with names to be somewhat complex :)
Short-hand seems to go nicely for quick prototyping with spaces in `BayesSearchCV`, but there actually one does not need named spaces, so yeah not sure whether we would need it in the end. Lets see what future users say.

Reply:
I agree we should try as much as possible to gently persuade people to not use the shorthand notation (anymore), for example by only allowing you to do simple stuff through it. Along the lines of `explicit is better than implicit`.

Reply:
I don't think gentle persuasion is my thing so I'll leave that to you! :-)

Seriously though, I think it would be a wise decision to aim for crystal clear semantics of the API rather than having a million clever little features. I can see both in your code and your github threads that you are also confused about your code and its semantics, and I'm glad that you have now realized the confusion is 10x for newcomers :-)